oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileStoreSplitter.java

```diff
@@ -0,0 +1,291 @@
+
+        return indexDefinitions;
+    }
+
+    public Set<NodeTypeInfo> getSplitNodeType(Set<IndexDefinition> indexDefinitions) throws IOException, CommitFailedException {
+        HashSet<String> nodeTypeNameSet = new HashSet<>();
+        Set<NodeTypeInfo> setOfNodeType = new HashSet<>();
+
+        for (IndexDefinition indexDf : indexDefinitions) {
+            Map<String, Aggregate> aggregateMap = indexDf.getAggregates();
+            nodeTypeNameSet.addAll(Objects.requireNonNull(aggregateMap).keySet());
+            nodeTypeNameSet.addAll(indexDf.getDefinedRules().stream().map(IndexDefinition.IndexingRule::getBaseNodeType).collect(Collectors.toList()));
+        }
+
+        for (String nodeTypeName : nodeTypeNameSet) {
+            setOfNodeType.add(infoProvider.getNodeTypeInfo(nodeTypeName));
+            setOfNodeType.addAll(getSubTypes(nodeTypeName));
+        }
+
+        log.debug("split node types: {}", setOfNodeType);
+        return setOfNodeType;
+    }
+
+    public Set<String> getPreferredPathElements(Set<IndexDefinition> indexDefinitions) {
+        Set<String> preferredPathElements = new HashSet<>();
+
+        for (IndexDefinition indexDf : indexDefinitions) {
+            preferredPathElements.addAll(indexDf.getRelativeNodeNames());
+        }
+        return preferredPathElements;
+    }
+
+    public List<FlatFileStore> split() throws IOException, CommitFailedException {
+        List<File> splitFlatFiles = new ArrayList<>();
+        List<FlatFileStore> splitFlatFileStores = new ArrayList<>();
+        try {
+            FileUtils.forceMkdir(workDir);
+        } catch (IOException e) {
+            log.error("failed to create split directory {}", workDir.getAbsolutePath());
+            splitFlatFileStores.add(ffs);
+            return splitFlatFileStores;
+        }
+
+        File originalFlatFile = new File(ffs.getFlatFileStorePath());
+        long fileSizeInBytes = useZip ? getGzipUncompressedSizeInBytes(ffs.getFlatFileStorePath()) : originalFlatFile.length();
+        log.info("original flatfile size: {}",  FileUtils.byteCountToDisplaySize(fileSizeInBytes));
+        long splitThreshold = Math.round((double) (fileSizeInBytes / splitSize));
+        log.info("split threshold: {} bytes, split size: {}",  FileUtils.byteCountToDisplaySize(splitThreshold), splitSize);
+
+
+        // return original if file too small or split size equals 1
+        if (splitThreshold < minimumSplitThreshold || splitSize <= 1) {
+            log.info("split is not necessary, skip splitting");
+            splitFlatFileStores.add(ffs);
+            return splitFlatFileStores;
+        }
+
+        Set<IndexDefinition> indexDefinitions = getIndexDefinitions();
+        Set<NodeTypeInfo> splitNodeTypes = getSplitNodeType(indexDefinitions);
+        Set<String>splitNodeTypesName = splitNodeTypes.stream().map(NodeTypeInfo::getNodeTypeName).collect(Collectors.toSet());
+        log.info("split allowed types: {}", splitNodeTypesName);
```

Comment:
@Ewocker a small change is preferred, can we merge above 3 lines code to the getSplitNodeTypeNames?

Reply:
updated